pkg/cdi/cache.go

```diff
@@ -65,17 +82,60 @@ func (c *Cache) Configure(options ...Option) error {
 	c.Lock()
 	defer c.Unlock()
 
+	return c.configure(options...)
+}
+
+// Configure the Cache. Start/stop CDI Spec directory watch, refresh
+// the Cache if necessary.
+func (c *Cache) configure(options ...Option) error {
+	var err error
+
 	for _, o := range options {
-		if err := o(c); err != nil {
+		if err = o(c); err != nil {
 			return errors.Wrapf(err, "failed to apply cache options")
 		}
 	}
 
+	c.dirErrors = make(map[string]error)
+
+	c.watch.stop()
+	if c.autoRefresh {
+		c.watch.setup(c.specDirs, c.dirErrors)
+		c.watch.start(&c.Mutex, c.refresh, c.dirErrors)
+	}
+	c.refresh()
+
 	return nil
 }
 
 // Refresh rescans the CDI Spec directories and refreshes the Cache.
+// In manual refresh mode the cache is always refreshed. In auto-
+// refresh mode the cache is only refreshed if it is out of date.
 func (c *Cache) Refresh() error {
+	c.Lock()
+	defer c.Unlock()
+
+	// collect cached errors globally as a single multierror
+	cachedErrors := func() error {
+		var result error
+		for _, err := range c.errors {
+			result = multierror.Append(result, err...)
+		}
+		return result
+	}
+
+	// We need to refresh if
+	// - we're in manual refresh mode, or
+	// - in auto-refresh mode a missing Spec dir appears (added to watch)
+	if !c.autoRefresh || c.watch.update(c.dirErrors) {
+		return c.refresh()
+	}
+
+	return cachedErrors()
```

Comment:
Does this mean that we lose access to the `cachedErrors` if we end up calling `refresh`? Is this intentional?

Reply:
No, it's sort of the opposite... we want to keep refresh idempotent with respect to its return value.

If we don't end up calling `refresh()` in case we're already up-to-date, then we still return the cached errors from the last `refresh()` to keep it idempotent. `cachedErrors()` is simply used as a macro here. It's a locally-scoped function pointer. Since it is only used in one place, if it is considered disturbing it can be replaced easily with the body of the anonymous function and the accompanied comment.